19 de febrero 2024.

Nancy Woolrich Sorroza

# Tarea - Programación lineal

### Diseño de la Dieta Óptima

Se quiere producir comida para gatos de la manera más barata, no obstante se debe también asegurar que se cumplan los datos requeridos de analisis nutricional. Por lo que se quiere variar la cantidad de cada ingrediente para cumplir con los estandares nutricionales. Los requisitos que se tienen es que en $100$ gramos, se deben tener **por lo menos** $8$ gramos de proteína y $6$ gramos de grasa. Así mismo, no se debe tener más de $2$ gramos de fibra y $0.4$ gramos de sal.  

Los datos nutricionales se pueden obtener de la siguiente tabla:

Ingrediente|Proteína|Grasa|Fibra|Sal
:----|----|----|----|----
Pollo|  10.0%|08.0%|00.1%|00.2%
Carne|  20.0%|10.0%|00.5%|00.5%
Cordero|15.0%|11.0%|00.5%|00.7%
Arroz|  00.0%|01.0%|10.0%|00.2%
Trigo|  04.0%|01.0%|15.0%|00.8%
Gel|    00.0%|00.0%|00.0%|00.0%

Los costos de cada producto son:

Ingrediente|Costo por gramo
:----|----
Pollo|$\$$0.013
Carne|$\$$0.008
Cordero|$\$$0.010
Arroz|$\$$0.002
Trigo|$\$$0.005
Gel|$\$$0.001    

Lo que se busca optimizar en este caso es la cantidad de productos que se debe utilizar en la comida de gato, minimizando el costo total. Para simplificar la notación use las siguientes variables: 

+ $x_1:$ Gramos de pollo  
+ $x_2:$ Gramos de carne  
+ $x_3:$ Gramos de cordero  
+ $x_4:$ Gramos de arroz  
+ $x_5:$ Gramos de trigo  
+ $x_6:$ Gramos de gel

La tarea consiste en plantear el problemade programación lineal que permita satisfacer las necesidades alimenticias del gato al tiempo que minimice el costo total y resolverlo con `linprog`.

### Criterio de optimización

- Minimizar el costo de producción de comida para gatos.

### Variables que puedo controlar
- Cantidad de cada ingrediente:

1. $x_1:$ Gramos de pollo  
2. $x_2:$ Gramos de carne  
3. $x_3:$ Gramos de cordero  
4. $x_4:$ Gramos de arroz  
5. $x_5:$ Gramos de trigo  
6. $x_6:$ Gramos de gel

### Función de costo

$$\min_{x_1,\dots,x_6} 0.013x_1 + 0.008x_2 + 0.010x_3 + 0.002x_4 + 0.005x_5 + 0.001x_6$$

### Restricciones

- $0.10x_1 + 0.20x_2 + 0.15x_3 + 0.02x_4 + 0.04x_5 + 0.00x_6\geq 8$

- $0.08x_1 + 0.10x_2 + 0.11x_3 + 0.01x_4 + 0.01x_5 + 0.00x_6\geq 6$

- $0.001x_1 + 0.005x_2 + 0.005x_3 + 0.10x_4 + 0.15x_5 + 0.00x_6\leq 2$

- $0.002x_1 + 0.005x_2 + 0.007x_3 + 0.002x_4 + 0.008x_5 + 0.00x_6\leq 4$

In [6]:
import numpy as np
import scipy.optimize as opt

c=np.array([-0.013,
            -0.008,
            -0.010,
            -0.002,
            -0.005,
            -0.001])

Aub=np.array([ 
    [-0.10, -0.20, -0.15, -0.02, -0.04, -0.00],
    [-0.08, -0.10, -0.11, -0.01, -0.01, -0.00],
    [0.001, 0.005, 0.005, 0.10, 0.15, 0.00],
    [0.002, 0.005, 0.007, 0.002, 0.008, 0.00]
])

bub=np.array([-8,
              -6,
              2, 
              4])

opt.linprog (c, Aub, bub)

       message: The problem is unbounded. (HiGHS Status 10: model_status is Unbounded; primal_status is At upper bound)
       success: False
        status: 3
           fun: None
             x: None
           nit: 2
         lower:  residual: None
                marginals: None
         upper:  residual: None
                marginals: None
         eqlin:  residual: None
                marginals: None
       ineqlin:  residual: None
                marginals: None

### Problema de distribución de energía eléctrica

La Comisión Federal de Electricidad **(CFE)** dispone de cuatro plantas de generación para satisfacer la demanda diaria eléctrica en cuatro ciudades, Guadalajara, León y Morelia. Las plantas $1$, $2$ y $3$ pueden satisfacer $80$, $40$ y $60$ millones de KW al día respectivamente. Las necesidades de las ciudades de Guadalajara, León y Morelia son de $70$, $40$ y $70$ millones de Kw al día respectivamente. 


Los costos asociados al envío de suministro energético por cada millón de Kw entre cada planta y cada ciudad son los registrados en la siguiente tabla. 

-|Guadalajara|León|Morelia
:----|----|----|----
Planta 1|5|2|7
Planta 2|3|6|6
Planta 3|6|1|2

Y por último, las restricciones del problema, van a estar dadas por las capacidades de oferta y demanda de cada planta (en millones de KW) y cada ciudad.

Para simplificar la notación use las siguientes variables:

+ $x_1$: Kw (en millones) distribuidos de la Planta 1 a Guadalajara
+ $x_2$: Kw (en millones) distribuidos de la Planta 1 a León
+ $x_3$: Kw (en millones) distribuidos de la Planta 1 a Morelia
+ $x_4$: Kw (en millones) distribuidos de la Planta 2 a Guadalajara
+ $x_5$: Kw (en millones) distribuidos de la Planta 2 a León
+ $x_6$: Kw (en millones) distribuidos de la Planta 2 a Morelia
+ $x_7$: Kw (en millones) distribuidos de la Planta 3 a Guadalajara
+ $x_8$: Kw (en millones) distribuidos de la Planta 3 a León
+ $x_9$: Kw (en millones) distribuidos de la Planta 3 a Morelia

La tarea consiste en plantear el problema de programación lineal que permita satisfacer las necesidades de todas las ciudades al tiempo que minimice los costos asociados a la distribución y resolverlo con `linprog`.

### Criterio de optimización

- Minimizar el costo de distribución de energía.

### Variables que puedo controlar
Cantidad de kw que se mandan por planta a cada ciudad:

+ $x_1$: Kw (en millones) distribuidos de la Planta 1 a Guadalajara
+ $x_2$: Kw (en millones) distribuidos de la Planta 1 a León
+ $x_3$: Kw (en millones) distribuidos de la Planta 1 a Morelia
+ $x_4$: Kw (en millones) distribuidos de la Planta 2 a Guadalajara
+ $x_5$: Kw (en millones) distribuidos de la Planta 2 a León
+ $x_6$: Kw (en millones) distribuidos de la Planta 2 a Morelia
+ $x_7$: Kw (en millones) distribuidos de la Planta 3 a Guadalajara
+ $x_8$: Kw (en millones) distribuidos de la Planta 3 a León
+ $x_9$: Kw (en millones) distribuidos de la Planta 3 a Morelia

### Función de costo

$$\min_{x_1,\dots,x_9} 5x_1 + 2x_2 + 7x_3 + 3x_4 + 6x_5 + 6x_6 + 6x_7 + x_8 + 2x_9$$

### Restricciones

- $70x_1 + 40x_2 + 70x_3 + 0x_4 + 0x_5 + 0x_6 + 0x_7 + 0x_8 + 0x_9\leq 80$

- $0x_1 +0x_2 + 0x_3 + 70x_4 + 40x_5 + 70x_6 + 0x_7 + 0x_8 + 0x_9\leq 40$

- $0x_1 + 0x_2 + 0x_3 + 0x_4 + 0x_5 + 0x_6 + 70x_7 + 40x_8 + 70x_9\leq 60$


In [3]:
import numpy as np
import scipy.optimize as opt

c=np.array([-5,
            -2,
            -7,
            -3,
            -6,
            -6,
            -6,
            -1,
            -2])

Aub=np.array([ 
    [70, 40, 70, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 70, 40, 70, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 70, 40, 70]
])

bub=np.array([80,
              40, 
              60])

opt.linprog (c, Aub, bub)

        message: Optimization terminated successfully. (HiGHS Status 7: Optimal)
        success: True
         status: 0
            fun: -19.142857142857142
              x: [ 0.000e+00  0.000e+00  1.143e+00  0.000e+00  1.000e+00
                   0.000e+00  8.571e-01  0.000e+00  0.000e+00]
            nit: 0
          lower:  residual: [ 0.000e+00  0.000e+00  1.143e+00  0.000e+00
                              1.000e+00  0.000e+00  8.571e-01  0.000e+00
                              0.000e+00]
                 marginals: [ 2.000e+00  2.000e+00  0.000e+00  7.500e+00
                              0.000e+00  4.500e+00  0.000e+00  2.429e+00
                              4.000e+00]
          upper:  residual: [       inf        inf        inf        inf
                                    inf        inf        inf        inf
                                    inf]
                 marginals: [ 0.000e+00  0.000e+00  0.000e+00  0.000e+00
                              0.000e+00  0.000e+00  

La **Planta 1** debe enviar **1.143 millones de kw** a **Morelia** para que el costo sea mínimo.

La **Planta 2** debe enviar **1 millón de kw** a **León** para que el costo sea mínimo.

La **Planta 3** debe enviar **0.8571 millones de kw** a **Guadalajara** para que el costo sea mínimo.